submariner/globalnet-enhancement2-0.md

```diff
@@ -0,0 +1,282 @@
+
+The following CRD is proposed to support the new design.
+
+```Go
+type GlobalnetEgressIP struct {
+    metav1.TypeMeta   `json:",inline"`
+    metav1.ObjectMeta `json:"metadata,omitempty"`
+
+    // Spec is the specification of desired behavior of GlobalnetEgressIP object.
+    Spec               EgressIPSpec `json:"spec"`
+
+    // Observed status of GlobalnetEgressIP. Its a read-only field.
+    // +optional
+    Status             EgressIPStatus `json:"status,omitempty"`
+}
+
+type EgressIPSpec struct {
+    // GlobalIPs is a list of EgressIPs requested. The requested globalIPs should be part
+    // of globalCIDR of the Cluster. If the requested globalIPs are not allocated to other
+    // K8s objects, then it will be allocated and the status field will be updated accordingly.
+    // If the globalIPs are already allocated to other objects, creation of the CRD will fail.
+    // If this field is not specified, a single globalIP will be allocated.
+    // This field cannot be changed through updates.
+    // +optional
+    GlobalIPs  []string `json:"globalIPs,omitempty"`
+
+    // Selects the Namespaces to which this GlobalnetEgressIP object applies.
+    // If an empty namespaceSelector: {} is configured, it selects all the Pods in all namespaces.
+    // On the other hand, if you omit specifying namespaceSelector, it does not select any
+    // namespaces and selects only Pods from the namespace where the GlobalnetEgressIP is
+    // deployed to.
+    // +optional
+    NamespaceSelector metav1.LabelSelector `json:"namespaceSelector,omitempty"`
+
+    // Selects the pods whose label match the definition. This is an optional field and
+    // in case it is not set, it results in all the Pods selected from the NamespaceSelector.
+    // In case it is set, its intersected with NamespaceSelector, and only Pods that match the
+    // PodSelector and present in the NamespaceSelector will have the GlobalIPs as EgressIPs.
+    // Either namespaceSelector or podSelector have to be specified. If both are omitted,
+    // its considered as an error.
+    // When a Pod or Namespace matches multiple GlobalnetEgressIP objects, there is no guarantee
+    // which of the globalIP address that are assigned to GlobalnetEgressIP will be used.
+    // +optional
+    PodSelector metav1.LabelSelector `json:"podSelector,omitempty"`
+}
+
+type EgressIPStatus struct {
+    // The list of GlobalIPs assigned via this GlobalnetEgressIP object.
+    GlobalIPs []string `json:"globalIPs"`
+
+    // The Namespaces to which the GlobalIPs are applied.
+    Namespaces []string `json:"namespaces"`
+
+    // The Pods to which the GlobalIPs are applied.
+    Pods []string `json:"pods"`
+}
+```
+
+Creation of GlobalnetEgressIP object will be controlled via a ClusterRole and only users having
+the necessary role will be able to create the objects. The GlobalnetEgressIP object can be
+created in any namespace with selectors that match any other namespaces.
+
+### User Stories
```

Comment:
Would older globalnet annotation still be supported? It's not clear

Reply:
Done.

Reply:
Can you provide examples for ingress IPs?

Reply:
The GlobalnetEgressIP CRD is for EgressIP only. 
By default, all exported Services (including Headless Services) will automatically get a globalIP (annotated on the Service object or ServiceExport object) and Globalnet will program rules to support **Ingress** traffic destined to the corresponding globalIP. 

Do you want me to add a use-case to show how/when Ingress globalIP is allocated to an exported service as shown below.

```
kubectl -n default create deployment nginx --image=nginxinc/nginx-unprivileged:stable-alpine
kubectl -n default expose deployment nginx --port=8080
subctl export service --namespace default nginx
```

Reply:
Yes I think it'll clarify it.
Also it should be clear that the only way to get an ingress IP is by exporting a service (clusterip of headless) and that users can't assign specific IPs.

Reply:
I thought it was clear from the name GlobalnetEgressIP :) I'll add it.